flatten.sh

```diff
@@ -0,0 +1,118 @@
+#! /bin/bash
+
+# Copyright (c) 2019, Arm Ltd.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+
+
+# This script comes with no warranty.
+
+# To run, just say:
+# ./flatten.sh
+
+# This takes the history of origin/master and creates branch called "new" with
+# the rewritten history. It takes about 20s on my laptop. The "new" branch has a
+# flat git history (that is, a series of commits with only one parent).
+# Flattening is done for merge commits by taking the content of the commit as it
+# is at the merge commit.
+
+# The commit message is preserved for mainline non-merge commits.
+
+# For merge commits, a new commit message is made by combining the commit
+# messages of the non-mainline commits.
+
+# Authorship (name, email, date) is taken from the second parent of the merge.
+
+# set -x
+set -e -u
+
+# First-order assumptions:
+#
+# * Pull requests are authored by one person.
+
+# Known deficiencies:
+#
+# * The commit messages are ugly. These can be tidied up fairly easily once a
+#   format is agreed.
+#
+# * Individual commits within PRs are lost. This is somewhat fundamental, and
+#   could be worked around with some effort; it would require rebasing a number
+#   of patches, which is risky and time consuming, so instead the working tree
+#   is taken as-is from the merge commit.
```

Comment:
I think this is a serious short-coming. Squashing together all of the commits in a pull request can lose a lot of information. See the new commit 5e2cd5eb066062b7bd6d7874817d5213187bccb2 for an example. It combines several commits, some of are refactorings and some are function changes.

Reply:
I agree that isn't great.

The only alternative I'm aware of is to do effectively an interactive rebase, which would also flatten history, discarding only the (otherwise empty) merge commits. As far as I can tell, that would require some effort to deal with various bits of the history where commits can't be replayed without the merge commit. I don't see why this couldn't be done in principle, but it would require additional engineering effort.

Reply:
@peterwaller-arm mentioned to me this evening that he might have a mitigation to this in that the new squashed commits could link back to the original F18 git shas that went into them. So the F18 git history in LLVM monorepo is straightline (and cheaply done) but has some commits that combine things together but with the original commits, that can be clicked through to in github, to help separate them.

Reply:
The next update does a better job. Now there are only 6 merge commits which get squashed. As described elsewhere, fixing those would require manual work. Please take another look.